### PR TITLE
Use Storage write API in BigQuery connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# BigQuery
* Improve performance when writing rows. ({issue}`18897`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java [hunk 1]
```diff
@@ -64,28 +73,46 @@ public BigQueryPageSink(
...
+                .map(tableName -> TableName.of(remoteTableName.getProjectId(), remoteTableName.getDatasetName(), tableName))
+                .orElseGet(remoteTableName::toTableName);
     }
 
     @Override
     public CompletableFuture<?> appendPage(Page page)
     {
-        InsertAllRequest.Builder batch = InsertAllRequest.newBuilder(tableId);
+        JSONArray batch = new JSONArray();
         for (int position = 0; position < page.getPositionCount(); position++) {
```
Comment (anusudarsan) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java:85]: should the data be batched based on a request size limit (config)? https://cloud.google.com/bigquery/quotas#write-api-limits

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java [hunk 2]
```diff
@@ -64,28 +73,46 @@ public BigQueryPageSink(
...
         }
 
-        client.insert(batch.build());
+        insertWithCommitted(batch);
         return NOT_BLOCKED;
     }
 
+    private void insertWithCommitted(JSONArray batch)
+    {
+        WriteStream stream = WriteStream.newBuilder().setType(COMMITTED).build();
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java:100]: we are creating a new stream per Page.
BigQuery has a limit of 1k streams open at a time and 30k/4 hours (7.5k creations per hour).

Let's either document this if we think it's shouldn't be a problem or let's consider creating a single stream per page-sink/tablewriter task.

Also do we consider to use "pending mode" long term to provide proper isolation? With current mode ("committed") if a single stream fails then writes from other streams will still succeed and be visible. i.e. it's not ACID and there's no way to rollback. Note that this is same behaviour that we already had so it's not a regression in that sense.

**Review by hashhar — APPROVED** [review]
looks good to me

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java (hunk 1 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java:89]: good to do but pre-existing issue.

File: plugin/trino-bigquery/pom.xml [hunk 3]
```diff
@@ -304,6 +309,12 @@
             <artifactId>httpcore</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20231013</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-bigquery/pom.xml:316]: Do you mean `com.google.cloud:libraries-bom`? The bom doesn't contain `org.json`
https://mvnrepository.com/artifact/com.google.cloud/libraries-bom/26.27.0

**Comment by hashhar** [comment]
@ebyhr Does the docs need updating about any new IAM permissions which are needed.

**Pull request merged**